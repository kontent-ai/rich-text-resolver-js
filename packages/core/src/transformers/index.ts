export * from "./html-transformer/index.js";
export * from "./portable-text-transformer/index.js";
export * from "./transformer-models.js";
