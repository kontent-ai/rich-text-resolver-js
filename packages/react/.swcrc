{
  "jsc": {
    "parser": {
      "syntax": "typescript",
      "tsx": true
    }
  },
  "module": {
    "type": "commonjs", 
    "strict": true
  },
  "env": {
    "targets": "> 0.25%, not dead",
    "mode": "usage",
    "coreJs": "3.37"
  }
} 