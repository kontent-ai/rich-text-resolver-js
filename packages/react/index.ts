export * from './src/react.js'; 