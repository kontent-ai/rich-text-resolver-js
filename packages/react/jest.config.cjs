const baseConfig = require('../../jest.config.base.cjs');

module.exports = {
  ...baseConfig,
  displayName: 'react',
  rootDir: '.',
  moduleNameMapper: {
    ...baseConfig.moduleNameMapper,
    '^@kontent-ai/rich-text-resolver$': '<rootDir>/../core/src',
    '^@kontent-ai/rich-text-resolver/(.*)$': '<rootDir>/../core/src/$1'
  }
}; 