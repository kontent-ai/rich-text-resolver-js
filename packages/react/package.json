{
    "name": "@kontent-ai/rich-text-resolver-react",
    "version": "3.0.0",
    "description": "React utilities for Kontent.ai rich text resolver",
    "main": "./dist/legacy/index.js",
    "module": "./dist/esnext/index.js",
    "browser": "./dist/legacy/index.js",
    "types": "./dist/esnext/index.d.ts",
    "type": "module",
    "repository": {
      "type": "git",
      "url": "https://github.com/kontent-ai/rich-text-resolver-js.git",
      "directory": "packages/react"
    },
    "exports": {
      ".": {
        "import": "./dist/esnext/index.js",
        "require": "./dist/legacy/index.js"
      }
    },
    "scripts": {
      "build:esnext": "tsc",
      "build:legacy": "swc src --out-dir dist/legacy && swc index.ts --out-file dist/legacy/index.js",
      "build": "npm run build:esnext && npm run build:legacy",
      "test": "jest",
      "lint": "eslint . --ext ts,tsx"
    },
    "peerDependencies": {
      "react": ">=18.0.0",
      "@kontent-ai/rich-text-resolver": "^3.0.0"
    },
    "dependencies": {
      "@portabletext/react": "^3.2.1",
      "@portabletext/types": "^2.0.13"
    },
    "devDependencies": {
      "@kontent-ai/delivery-sdk": "^16.0.0",
      "@kontent-ai/rich-text-resolver": "*",
      "@testing-library/react": "^16.3.0",
      "@types/react": "^19.1.0",
      "react": "^19.1.0"
    },
    "publishConfig": {
      "access": "public",
      "registry": "https://registry.npmjs.org/"
    }
  }