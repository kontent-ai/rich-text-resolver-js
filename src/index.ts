export { toVueImageDefault } from './utils/resolution/vue.js';
export { toHTMLImageDefault } from './utils/resolution/html.js';
export { traversePortableText } from './utils/transformer-utils.js';
export { transformToJson } from "./transformers/json-transformer/json-transformer.js";
export { transformToPortableText } from "./transformers/portable-text-transformer/portable-text-transformer.js";
export { parse as browserParse } from "./parser/browser/rich-text-browser-parser.js";
export { parse as nodeParse } from "./parser/node/rich-text-node-parser.js";
