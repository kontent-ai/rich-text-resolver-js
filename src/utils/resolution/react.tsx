import {
  PortableTextMarkComponent,
  PortableTextReactComponents,
  PortableTextTypeComponent,
  toPlainText,
} from "@portabletext/react";
import React from "react";

import {
  PortableTextComponentOrItem,
  PortableTextExternalLink,
  PortableTextImage,
  PortableTextItemLink,
  PortableTextTable,
  PortableTextTableCell,
  PortableTextTableRow,
} from "../../transformers/transformer-models.js";

type RichTextCustomBlocks = {
  image: PortableTextTypeComponent<PortableTextImage>;
  componentOrItem: PortableTextTypeComponent<PortableTextComponentOrItem>;
  table: PortableTextTypeComponent<PortableTextTable>;
};

type RichTextCustomMarks = {
  contentItemLink: PortableTextMarkComponent<PortableTextItemLink>;
  link: PortableTextMarkComponent<PortableTextExternalLink>;
};

/**
 * Extends `PortableTextReactComponents` type from `@portabletext/react` package with
 * pre-defined types for resolution of Kontent.ai specific custom objects.
 */
export type PortableTextReactResolvers = Partial<
  Omit<PortableTextReactComponents, "types" | "marks"> & {
    types: PortableTextReactComponents["types"] & Partial<RichTextCustomBlocks>;
    marks: PortableTextReactComponents["marks"] & Partial<RichTextCustomMarks>;
  }
>;

export const TableComponent: React.FC<PortableTextTable> = (table) => {
  const renderCell = (cell: PortableTextTableCell, cellIndex: number) => {
    return <td key={cellIndex}>{toPlainText(cell.content)}</td>;
  };

  const renderRow = (row: PortableTextTableRow, rowIndex: number) => {
    return <tr key={rowIndex}>{row.cells.map(renderCell)}</tr>;
  };

  return (
    <table>
      <tbody>{table.rows.map(renderRow)}</tbody>
    </table>
  );
};

export const ImageComponent: React.FC<PortableTextImage> = (image) => (
  <img src={image.asset.url} alt={image.asset.alt ?? ""} />
);
