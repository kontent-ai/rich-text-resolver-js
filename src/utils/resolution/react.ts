import { PortableTextMarkComponent, PortableTextReactComponents, PortableTextTypeComponent } from "@portabletext/react";

import {
  PortableTextComponentOrItem,
  PortableTextExternalLink,
  PortableTextImage,
  PortableTextItemLink,
  PortableTextTable,
} from "../../transformers/transformer-models.js";

type RichTextCustomBlocks = {
  image: PortableTextTypeComponent<PortableTextImage>;
  componentOrItem: PortableTextTypeComponent<PortableTextComponentOrItem>;
  table: PortableTextTypeComponent<PortableTextTable>;
};

type RichTextCustomMarks = {
  contentItemLink: PortableTextMarkComponent<PortableTextItemLink>;
  link: PortableTextMarkComponent<PortableTextExternalLink>;
};

/**
 * Extends `PortableTextReactComponents` type from `@portabletext/react` package with
 * pre-defined types for resolution of Kontent.ai specific custom objects.
 */
export type PortableTextReactResolvers = Partial<
  Omit<
    PortableTextReactComponents,
    "types" | "marks"
  > & {
    types: PortableTextReactComponents["types"] & Partial<RichTextCustomBlocks>;
    marks: PortableTextReactComponents["marks"] & Partial<RichTextCustomMarks>;
  }
>;
