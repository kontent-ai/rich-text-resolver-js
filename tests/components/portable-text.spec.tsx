import { Elements, ElementType } from "@kontent-ai/delivery-sdk";
import { PortableText } from "@portabletext/react";
import { render } from "@testing-library/react";
import React from "react";

import { transformToPortableText } from "../../src";
import { ImageComponent, PortableTextReactResolvers, TableComponent } from "../../src/utils/resolution/react";

const dummyRichText: Elements.RichTextElement = {
  value: "<p>some text in a paragraph</p>",
  type: ElementType.RichText,
  images: [],
  linkedItemCodenames: [],
  linkedItems: [
    {
      system: {
        id: "99e17fe7-a215-400d-813a-dc3608ee0294",
        name: "test item",
        codename: "test_item",
        language: "default",
        type: "test",
        collection: "default",
        sitemapLocations: [],
        lastModified: "2022-10-11T11:27:25.4033512Z",
        workflowStep: "published",
        workflow: "default",
      },
      elements: {
        text_element: {
          type: ElementType.Text,
          name: "text element",
          value: "random text value",
        },
      },
    },
  ],
  links: [],
  name: "dummy",
};

const portableTextComponents: PortableTextReactResolvers = {
  types: {
    componentOrItem: ({ value }) => {
      const item = dummyRichText.linkedItems.find(item => item.system.codename === value.component._ref);
      return <div>{item?.elements.text_element.value}</div>;
    },
    table: ({ value }) => <TableComponent {...value} />,
    image: ({ value }) => <ImageComponent {...value} />,
  },
  marks: {
    link: ({ value, children }) => {
      const target = (value?.href || "").startsWith("http") ? "_blank" : undefined;
      return (
        <a
          href={value?.href}
          target={target}
          rel={value?.rel}
          title={value?.title}
          data-new-window={value?.["data-new-window"]}
        >
          {children}
        </a>
      );
    },
    contentItemLink: ({ value, children }) => {
      const item = dummyRichText.linkedItems.find(item => item.system.id === value?.reference._ref);
      return (
        <a href={"https://somerandomwebsite.xyz/" + item?.system.codename}>
          {children}
        </a>
      );
    },
  },
};

describe("portable text React resolver", () => {
  const renderPortableText = (richTextValue: string, components = portableTextComponents) => {
    dummyRichText.value = richTextValue;
    const portableText = transformToPortableText(dummyRichText.value);

    return render(<PortableText value={portableText} components={components} />).container.innerHTML;
  };

  it("renders simple HTML", () => {
    const tree = renderPortableText("<p>some text in a paragraph</p>");
    expect(tree).toMatchSnapshot();
  });

  it("renders a resolved linked item", () => {
    const tree = renderPortableText(
      "<object type=\"application/kenticocloud\" data-type=\"item\" data-rel=\"link\" data-codename=\"test_item\"></object>",
    );
    expect(tree).toMatchSnapshot();
  });

  it("renders internal and external links", () => {
    const tree = renderPortableText(`
      <p>
        <a href="http://google.com" title="linktitle" target="_blank">external link</a>
        <a data-item-id="99e17fe7-a215-400d-813a-dc3608ee0294" href=""><strong>item link</strong></a>
      </p>
    `);
    expect(tree).toMatchSnapshot();
  });

  it("renders a table", () => {
    const tree = renderPortableText(`
      <table><tbody>
        <tr><td>Ivan</td><td>Jiri</td></tr>
        <tr><td>Ondra</td><td>Dan</td></tr>
      </tbody></table>
    `);
    expect(tree).toMatchSnapshot();
  });

  it("renders an image", () => {
    const tree = renderPortableText(`
      <p>some text before an asset</p>
      <figure data-asset-id="bc6f3ce5-935d-4446-82d4-ce77436dd412" data-image-id="bc6f3ce5-935d-4446-82d4-ce77436dd412">
        <img src="https://assets-us-01.kc-usercontent.com:443/cec32064-07dd-00ff-2101-5bde13c9e30c/7d534724-edb8-4a6d-92f6-feb52be61d37/image1_w_metadata.jpg" data-asset-id="bc6f3ce5-935d-4446-82d4-ce77436dd412" data-image-id="bc6f3ce5-935d-4446-82d4-ce77436dd412" alt="" />
      </figure>
    `);
    expect(tree).toMatchSnapshot();
  });
});
