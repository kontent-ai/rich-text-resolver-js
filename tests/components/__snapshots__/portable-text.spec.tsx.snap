// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`portable text React resolver renders a resolved linked item 1`] = `"<div>random text value</div>"`;

exports[`portable text React resolver renders a table 1`] = `"<table><tbody><tr><td>Ivan</td><td>Jiri</td></tr><tr><td>Ondra</td><td>Dan</td></tr></tbody></table>"`;

exports[`portable text React resolver renders an image 1`] = `"<p>some text before an asset</p><img alt="" src="https://assets-us-01.kc-usercontent.com:443/cec32064-07dd-00ff-2101-5bde13c9e30c/7d534724-edb8-4a6d-92f6-feb52be61d37/image1_w_metadata.jpg">"`;

exports[`portable text React resolver renders internal and external links 1`] = `"<p><a href="http://google.com" target="_blank" title="linktitle">external link</a><a href="https://somerandomwebsite.xyz/test_item"><strong>item link</strong></a></p>"`;

exports[`portable text React resolver renders simple HTML 1`] = `"<p>some text in a paragraph</p>"`;
